llama-cpp-2/src/sampling.rs

```diff
@@ -21,236 +21,155 @@ impl Debug for LlamaSampler {
     }
 }
 
-impl LlamaSampler {
-    /// Create a new `LlamaSampler` from the given parameters.
-    /// # Errors
-    /// Returns an error if the underlying C++ code returns a null pointer.
-    pub fn new(params: params::LlamaSamplerChainParams) -> Result<Self, LlamaSamplerError> {
-        let sampler = unsafe {
-            NonNull::new(llama_cpp_sys_2::llama_sampler_chain_init(
-                params.sampler_chain_params,
-            ))
-            .ok_or(LlamaSamplerError::NullReturn)
-        }?;
-
-        Ok(Self { sampler })
-    }
-
-    /// Samples the token with the largest probability.
-    #[must_use]
-    #[allow(unused_mut)]
-    pub fn add_greedy(mut self) -> Self {
-        unsafe {
-            let greedy_sampler = llama_cpp_sys_2::llama_sampler_init_greedy();
-            llama_cpp_sys_2::llama_sampler_chain_add(self.sampler.as_ptr(), greedy_sampler);
-        }
-
-        self
-    }
-
-    /// Samples according to the probability distribution of the tokens.
-    #[must_use]
-    #[allow(unused_mut)]
-    pub fn add_dist(mut self, seed: u32) -> Self {
-        unsafe {
-            let dist_sampler = llama_cpp_sys_2::llama_sampler_init_dist(seed);
-            llama_cpp_sys_2::llama_sampler_chain_add(self.sampler.as_ptr(), dist_sampler);
-        }
-
-        self
-    }
-
-    /// Top-K sampling described in academic paper "The Curious Case of Neural Text Degeneration" <https://arxiv.org/abs/1904.09751>
-    #[must_use]
-    #[allow(unused_mut)]
-    pub fn add_top_k(mut self, k: i32) -> Self {
-        unsafe {
-            let top_k_sampler = llama_cpp_sys_2::llama_sampler_init_top_k(k);
-            llama_cpp_sys_2::llama_sampler_chain_add(self.sampler.as_ptr(), top_k_sampler);
-        }
-
-        self
-    }
-
-    /// Nucleus sampling described in academic paper "The Curious Case of Neural Text Degeneration" <https://arxiv.org/abs/1904.09751>
-    #[must_use]
-    #[allow(unused_mut)]
-    pub fn add_top_p(mut self, p: f32, min_keep: usize) -> Self {
-        unsafe {
-            let top_p_sampler = llama_cpp_sys_2::llama_sampler_init_top_p(p, min_keep);
-            llama_cpp_sys_2::llama_sampler_chain_add(self.sampler.as_ptr(), top_p_sampler);
-        }
-
-        self
-    }
-
-    /// Minimum P sampling as described in <https://github.com/ggerganov/llama.cpp/pull/3841>
-    #[must_use]
-    #[allow(unused_mut)]
-    pub fn add_min_p(mut self, p: f32, min_keep: usize) -> Self {
-        unsafe {
-            let min_p_sampler = llama_cpp_sys_2::llama_sampler_init_min_p(p, min_keep);
-            llama_cpp_sys_2::llama_sampler_chain_add(self.sampler.as_ptr(), min_p_sampler);
-        }
+unsafe fn new_inner(params: LlamaSamplerParams) -> *mut llama_cpp_sys_2::llama_sampler {
```

Comment:
I don't think this has to be marked as unsafe. Are there invariants callers need to uphold? If so, document them.